numpy/core/src/multiarray/number.c

```diff
@@ -814,213 +814,107 @@ _array_nonzero(PyArrayObject *mp)
-        return NULL;
-    }
-    if (Py_TYPE(pv)->tp_as_number == 0) {
-        PyErr_SetString(PyExc_TypeError, "cannot convert to a "\
-                        "float; scalar object is not a number");
-        Py_DECREF(pv);
-        return NULL;
-    }
-    if (Py_TYPE(pv)->tp_as_number->nb_float == 0) {
-        PyErr_SetString(PyExc_TypeError, "don't know how to convert "\
-                        "scalar number to float");
-        Py_DECREF(pv);
-        return NULL;
-    }
-    /*
-     * If we still got an array which can hold references, stop
-     * because it could point back at 'v'.
-     */
-    if (PyArray_Check(pv) &&
-                    PyDataType_REFCHK(PyArray_DESCR((PyArrayObject *)pv))) {
-        PyErr_SetString(PyExc_TypeError,
-                "object array may be self-referencing");
-        Py_DECREF(pv);
-        return NULL;
-    }
-    pv2 = Py_TYPE(pv)->tp_as_number->nb_float(pv);
-    Py_DECREF(pv);
-    return pv2;
+    return array_scalar_forward(v, &PyNumber_Float, "in ndarray.__float__");
 }
 
-#if !defined(NPY_PY3K)
+#if defined(NPY_PY3K)
 
-static PyObject *
+NPY_NO_EXPORT PyObject *
+array_int(PyArrayObject *v)
+{
+    return array_scalar_forward(v, &PyNumber_Long, "in ndarray.__int__");
+}
+
+#else
+
+NPY_NO_EXPORT PyObject *
+array_int(PyArrayObject *v)
+{
+    return array_scalar_forward(v, &PyNumber_Int, "in ndarray.__int__");
+}
+
+NPY_NO_EXPORT PyObject *
 array_long(PyArrayObject *v)
 {
-    PyObject *pv, *pv2;
-    if (PyArray_SIZE(v) != 1) {
-        PyErr_SetString(PyExc_TypeError, "only length-1 arrays can "\
-                        "be converted to Python scalars");
-        return NULL;
-    }
-    pv = PyArray_GETITEM(v, PyArray_DATA(v));
-    if (pv == NULL) {
-        return NULL;
-    }
-    if (Py_TYPE(pv)->tp_as_number == 0) {
-        PyErr_SetString(PyExc_TypeError, "cannot convert to an int; "\
-                        "scalar object is not a number");
-        Py_DECREF(pv);
-        return NULL;
-    }
-    if (Py_TYPE(pv)->tp_as_number->nb_long == 0) {
-        PyErr_SetString(PyExc_TypeError, "don't know how to convert "\
-                        "scalar number to long");
-        Py_DECREF(pv);
+    return array_scalar_forward(v, &PyNumber_Long, "in ndarray.__long__");
+}
+
+/* hex and oct aren't exposed to the C api, but we need a function pointer */
+static PyObject *
+_PyNumber_Oct(PyObject *o) {
+    PyObject *res;
+    PyObject *mod = PyImport_ImportModule("builtins");
+    if (mod != NULL) {
         return NULL;
     }
-    /*
-     * If we still got an array which can hold references, stop
-     * because it could point back at 'v'.
-     */
-    if (PyArray_Check(pv) &&
-                    PyDataType_REFCHK(PyArray_DESCR((PyArrayObject *)pv))) {
-        PyErr_SetString(PyExc_TypeError,
-                "object array may be self-referencing");
-        Py_DECREF(pv);
+    res = PyObject_CallMethodObjArgs(mod, "oct", o, NULL);
+    Py_DECREF(module);
```

Comment:
"module" is not defined, you probably meant "mod"

Reply:
Good catch, only tested on python 3 locally